### PR TITLE
[Improvement] Training custom classes of ava dataset

### PR DESCRIPTION
**Pull request by irvingzhang0512**
## Target
Training some of the 80 ava classes to save training time and hopefully get better results for selected classes.

## TODO
+ [x] dataset/evaluation codes.
+ [x] unittest
+ [x] docs
+ [x] sample config
+ [x] model zoo, compare results.
+ [x] Add input arg `topk` for `BBoxHeadAVA`, because `num_classes` may be smaller than 5.
+ [x]  ~check whether `exclude_file_xxx` will affect the results.~

## results
+ slowonly_kinetics_pretrained_r50_4*16

|custom classes|mAP(train 80 classes)|mAP (train custom classes only)|selected classes comment|
|:-:|-:|-:|-:|
|range(1, 15)|**0.3460**|0.3399|all `PERSON_MOVEMENT` classes|
|`[11, 12, 14, 15, 79, 80]`|**0.7066**|0.7011|AP(80 classes ckpt) > 0.6|
|`[1,4,8,9,13,17,28,49,74]`|0.4339|**0.4397**|AP(80 classes ckpt) in[0.3, 0.6)|
|`[3, 6, 10, 27, 29, 38, 41, 48, 51, 53, 54, 59, 61, 64, 70, 72]`|0.1948|**0.3311**|AP(80 classes ckpt) in[0.1, 0.3)|
|`[11,12,17,74,79,80]`|**0.6520**|0.6438|> 50000 samples|
|`[1,8,14,59]`|0.4307|**0.5549**|[5000, 50000) samples|
|`[3,4,6,9,10,15,27,28,29,38,41,48,49,54,61,64,65,66,67,70,77]`|0.2384|**0.3269**|[1000, 5000) samples|
|`[22,37,47,51,63,68,72,78]`|0.0753|**0.3209**|[500, 1000) samples|
|`[2,5,7,13,20,24,26,30,34,36,42,45,46,52,56,57,58,60,62,69,73,75,76]`|0.0348|**0.1806**|[100, 500) samples|
|`[16,18,19,21,23,25,31,32,33,35,39,40,43,44,50,53,55,71]`|0.0169|**0.1984**|<100  samples|

## insights

I think ava dataset suffers from series class imbalance. Training custom classes helps to get better results for classes with fewer samples.

image

### CODE REVIEW

File: configs/detection/ava/README.md [hunk 1]
```diff
@@ -80,6 +80,18 @@ python tools/train.py configs/detection/AVA/slowonly_kinetics_pretrained_r50_8x8
 
 For more details and optional arguments infos, you can refer to **Training setting** part in [getting_started](/docs/getting_started.md#training-setting) .
 
+### Train Custom Classes From Ava Dataset
```
Comment (kennymckormick) [inline configs/detection/ava/README.md:83]: Can also add a trained ckpt, introduce the classes you use (better have some reasons), and add the performance.

**Comment by kennymckormick** [comment]
Some minor comments

**Comment by irvingzhang0512** [comment]
Maybe this pr is not that useful as I expected... Training custom classes leads to lower mAP...

+ slowonly_kinetics_pretrained_r50_4*16

|custom classes|mAP(trained 80 classes)|mAP (training custom classes only)|comment|
|:-:|-:|-:|-:|
|range(1, 15)|0.3460|0.3399|training all `PERSON_MOVEMENT` classes|

I'll do more tests
+ select classes with fewer samples only
+ finetune from 80 classes ckpt
+ another backbone
+ tune lr

@innerlee @kennymckormick any suggestions?

**Comment by kennymckormick** [comment]
> Maybe this pr is not that useful as I expected... Training custom classes leads to lower mAP... Most classes with hight AP(>0.3) get worse results.
> 
> * slowonly_kinetics_pretrained_r50_4*16
> 
> custom classes	mAP(trained 80 classes)	mAP (training custom classes only)	comment
> range(1, 15)	0.3460	0.3399	training all `PERSON_MOVEMENT` classes
> I'll do more tests
> 
> * select classes with fewer samples only
> * finetune from 80 classes ckpt
> * another backbone
> * tune lr
> 
> @innerlee @kennymckormick any suggestions?

Maybe you can try to pick the best model among epoch 11 -> 20, and see if its performance is better.

**Comment by irvingzhang0512** [comment]
> > Maybe this pr is not that useful as I expected... Training custom classes leads to lower mAP... Most classes with hight AP(>0.3) get worse results.
> > 
> > * slowonly_kinetics_pretrained_r50_4*16
> > 
> > custom classes	mAP(trained 80 classes)	mAP (training custom classes only)	comment
> > range(1, 15)	0.3460	0.3399	training all `PERSON_MOVEMENT` classes
> > I'll do more tests
> > 
> > * select classes with fewer samples only
> > * finetune from 80 classes ckpt
> > * another backbone
> > * tune lr
> > 
> > @innerlee @kennymckormick any suggestions?
> 
> Maybe you can try to pick the best model among epoch 11 -> 20, and see if its performance is better.

Actually, I choose the best ckpt(epoch no.14), not the last one..

**Comment by irvingzhang0512** [comment]
Finally get some good results... 


+ slowonly_kinetics_pretrained_r50_4*16

|custom classes|mAP(train 80 classes)|mAP (train custom classes only)|comment|
|:-:|-:|-:|-:|
|range(1, 15)|0.3460|0.3399|training all `PERSON_MOVEMENT` classes|
|`[11, 12, 14, 15, 79, 80]`|0.7066|0.7011|classes with mAP(80 classes ckpt) > 0.6|
|`[1, 8, 9, 13, 17, 28, 49, 74]`|0.4243|0.4459|classes with mAP(80 classes ckpt) in[0.3, 0.6)|
|`[3, 6, 10, 27, 29, 38, 41, 48, 51, 53, 54, 59, 61, 64, 70, 72]`|0.1948|0.3311|classes with mAP(80 classes ckpt) in[0.1, 0.3)|

+ I think ava suffers from data imbalance issue.  For slowonly_kinetics_pretrained_r50_4*16 ava model

|class|cnt of samples|AP|
|:-:|:-:|:-:|
|stand|164932|0.7833|
|sit|98770|0.7709|
|answer phone|3251|0.7082|
|talk to (e.g., self, a person, a group)|107742|0.6987|
|walk|40472|0.6434|
|watch (a person)|157920|0.6354|
|drive (e.g., a car, a truck)|1174|0.5321|
|carry/hold (an object)|78144|0.5127|
|listen to (a person)|103358|0.5107|
|lie/sleep|5298|0.4779|
|dance|3196|0.4612|
|martial art|2068|0.3788|
|swim|107|0.3687|
|bend/bow (at the waist)|8148|0.3376|
|ride (e.g., a bike, a car, a horse)|4663|0.3256|
|fight/hit (a person)|2602|0.2951|
|touch (an object)|15981|0.2639|
|watch (e.g., TV)|1917|0.2615|
|kiss (a person)|697|0.2456|
|run/jog|3296|0.2435|
|read|2048|0.2407|
|eat|2868|0.2276|
|drink|2194|0.1937|
|shovel|65|0.182|
|play musical instrument|1286|0.1776|
|crouch/kneel|2196|0.1603|
|hug (a person)|1043|0.1566|
|open (e.g., a window, a car door)|1228|0.1414|
|smoke|2855|0.1137|
|get up|1113|0.1129|
|sail boat|634|0.1005|
|close (e.g., a door, a box)|778|0.0995|
|hand clap|1181|0.0989|
|sing to (e.g., self, a person, a group)|1548|0.0802|
|fall down|290|0.0753|
|fishing|77|0.0572|
|grab (a person)|1866|0.0536|
|work on a computer|172|0.0529|
|write|749|0.0526|
|hand shake|604|0.0442|
|give/serve (an object) to (a person)|1008|0.0423|
|cut|174|0.0394|
|shoot|221|0.0364|
|text on/look at a cellphone|393|0.0349|
|dig|38|0.0328|
|enter|217|0.0314|
|put down|528|0.0278|
|take (an object) from (a person)|598|0.0243|
|row boat|62|0.0239|
|lift (a person)|376|0.0223|
|jump/leap|241|0.0189|
|dress/put on clothing|313|0.0174|
|lift/pick up|497|0.0168|
|push (another person)|350|0.0156|
|push (an object)|378|0.0139|
|climb (e.g., a mountain)|244|0.013|
|crawl|150|0.0125|
|listen (e.g., to music)|593|0.0082|
|pull (an object)|329|0.0078|
|throw|247|0.0077|
|hand wave|327|0.0068|
|take a photo|178|0.0023|
|turn (e.g., a screwdriver)|133|0.0023|
|press|87|0.0017|
|cook|61|0.0015|
|hit (an object)|167|0.0015|
|play with pets|121|0.0015|
|exit|53|0.0009|
|play with kids|122|0.0009|
|stir|54|0.0007|
|play board game|92|0.0006|
|point to (an object)|96|0.0006|
|clink glass|52|0.0004|
|kick (a person)|49|0.0004|
|chop|36|0.0003|
|brush teeth|18|0.0002|
|catch (an object)|80|0.0002|
|extract|7|0.0002|
|kick (an object)|34|0.0002|
|paint|63|0.0001|

+ We could train custom classes with fewer samples to get better results.

**Comment by kennymckormick** [comment]
Good results!  You can choose a successful example to add configs and checkpoints. Also remember to add ur results to README file.

**Comment by irvingzhang0512** [comment]
ckpt/json/log are uploaded [here](https://wp.cethik.com:443/link/D7DBF461E68BE62DD3B9DB9895EA71A8)
@kennymckormick Would you kindly help to upload these files to your server?

**Comment by irvingzhang0512** [comment]
kindly ping @kennymckormick

**Comment by kennymckormick** [comment]
> kindly ping @kennymckormick

Got it

**Comment by irvingzhang0512** [comment]
Add ckpt trained by custom classes in model zoo along with other models may be a little bit misleading. How about create a table for custom classes only...

**Comment by kennymckormick** [comment]
> ckpt/json/log are uploaded [here](https://wp.cethik.com:443/link/D7DBF461E68BE62DD3B9DB9895EA71A8)
> @kennymckormick Would you kindly help to upload these files to your server?

Sorry seems I have no permission to download the files. Would you please upload the ckpts with other tools? (like dropbox, googledrive, baiduyun or transfer.sh)

**Comment by irvingzhang0512** [comment]
> > ckpt/json/log are uploaded [here](https://wp.cethik.com:443/link/D7DBF461E68BE62DD3B9DB9895EA71A8)
> > @kennymckormick Would you kindly help to upload these files to your server?
> 
> Sorry seems I have no permission to download the files. Would you please upload the ckpts with other tools? (like dropbox, googledrive, baiduyun or transfer.sh)

I'll upload these to baiduyun tonight... Cannot use other tools while in the company.

**Comment by kennymckormick** [comment]
> Add ckpt trained by custom classes in model zoo along with other models may be a little bit misleading. How about create a table for custom classes only...

I think we need a pair of checkpoint and config file for users to try this feature. Besides, I think we can create a separate table for custom classes.

**Comment by irvingzhang0512** [comment]
@kennymckormick 
链接：https://pan.baidu.com/s/1XLL3yOqFr1IWM2eX5_nNGQ 
提取码：os3p

**Comment by kennymckormick** [comment]
> @kennymckormick
> 链接：https://pan.baidu.com/s/1XLL3yOqFr1IWM2eX5_nNGQ
> 提取码：os3p

gotcha

**Comment by kennymckormick** [comment]
@innerlee  ready now

**Comment by kennymckormick** [comment]
Would u plz help solve the lint issue @irvingzhang0512 , I did't install markdown lint on the machine I commit

**Review by kennymckormick — APPROVED** [review]

**Comment by kennymckormick** [comment]
@innerlee

File: configs/detection/ava/README.md [hunk 2]
```diff
@@ -80,6 +80,26 @@ python tools/train.py configs/detection/AVA/slowonly_kinetics_pretrained_r50_8x8
...
+
+You can train custom classes from ava. Ava suffers from class imbalance. There are more then 100,000 samples for classes like `stand`/`listen to (a person)`/`talk to (e.g., self, a person, a group)`/`watch (a person)`, whereas half of all classes has less than 500 samples. In most cases, training custom classes with fewer samples only will lead to better results.
+
+Three steps to train custom classes:
+
+- Step 1: Select custom classes from original classes, named `custom_classes`. Class `0` should not be selected since it is designed for potential usage(not implemented yet) and will be added automatically.
```
Comment (innerlee) [inline configs/detection/ava/README.md:89]: What does `potential usage` mean?

btw space before opening bracket `(`

Comment (irvingzhang0512) [reply configs/detection/ava/README.md:89]: Actually, I saw @kennymckormick mentioned `potential usage` somewhere...

Comment (irvingzhang0512) [reply configs/detection/ava/README.md:89]: https://github.com/open-mmlab/mmaction2/issues/508#issuecomment-753738568

I'll add the message here.

**Comment by innerlee** [comment]
Sorry I need to get a understanding on the motivation. Does custom==subset here? What if you simply filter the annotation file? Or use some re-sampling strategy to solve the class imbalance problem?

**Comment by kennymckormick** [comment]
> Sorry I need to get a understanding on the motivation. Does custom==subset here? What if you simply filter the annotation file? Or use some re-sampling strategy to solve the class imbalance problem?

Here custom==subset. The motivation is: 1. Sometimes you only need to recognize a subset of action classes, and training on the subset led to better performance. 2. Adding a line in the config file is better than regenerating annotation files when doing filtering.

**Comment by innerlee** [comment]
So an alternative implementation would be, writing a scirpt to fitler a subset of the annotation file, right?

An even better alternative is, implementing a sampling strategy on a class-basis, so that you can adjust the sampling ratio of any class to any propotions, including exclusion by setting `0` ratio?

**Comment by kennymckormick** [comment]
> So an alternative implementation would be, writing a scirpt to fitler a subset of the annotation file, right?
> 
> An even better alternative is, implementing a sampling strategy on a class-basis, so that you can adjust the sampling ratio of any class to any propotions, including exclusion by setting `0` ratio?

The first alternative is OK, but sometimes you don't want to have additional files generated.
For the second alternative, actually it is already implemented in MMAction2 (to apply it to AVA, a little modification to ava_dataset.py might be needed). Maybe we can discuss about which is the better design.

**Comment by irvingzhang0512** [comment]
> So an alternative implementation would be, writing a scirpt to fitler a subset of the annotation file, right?
> 
> An even better alternative is, implementing a sampling strategy on a class-basis, so that you can adjust the sampling ratio of any class to any propotions, including exclusion by setting `0` ratio?

There is a little difference between the first alternative and this pr. This pr modifies label id and num_classes. I was planning to train ava (or my own dataset) with softmax loss (single label instead of multi label) someday.

**Comment by innerlee** [comment]
I will post a discussion later today :)

**Comment by irvingzhang0512** [comment]
kindly ping @innerlee

**Comment by innerlee** [comment]
I'm still trying to understand the motivation, so please be patient.

* Q1: In the comparision tables, there are `mAP(train 80 classes)` and `mAP (train custom classes only)`. Are they evaluated using the same amount of samples?
* Q2: Does each row in the table correspond to a new model that is trained using the custom class in the row?

**Comment by irvingzhang0512** [comment]
> * Q1: In the comparision tables, there are `mAP(train 80 classes)` and `mAP (train custom classes only)`. Are they evaluated using the same amount of samples?

Yes

> * Q2: Does each row in the table correspond to a new model that is trained using the custom class in the row?

Yes

> * Q3: Maybe this has been asked before, what if you simply remove the records that do not belong to custom classes in the original `ava_action_list_v2.1.pbtxt`? Or rephrase as this: can this be implemented by simply creating a new dataset list?

We should remove records that do not belong to custom classes **AND change the label id of the samples that belong to custom classes**. Sorry about my poor english, I may explain in chinese here.

为什么要对custom classes重新编号而不是使用原始编号？
1) 如果使用ava的默认配置（multi-label任务，训练每个class使用sigmoid），那改变不改变标签编号对性能并没有影响（因为使用了sigmoid），但inference time和model size能省一点算一点。
2) 考虑到sit/stand/walk/lie这些是互斥的，后续可能会支持单标签训练（softmax），可能会得到更好的效果。使用softmax的时候，我们**猜测**num_classes会影响结果，假设custom_classes只有10类，AVA本身是80类，那最终的预测结果（softmax后）的shape应该是`[batch size, 10]`而不是`[batch size, 80]`，所以需要修改label id。当然这个需要以后试验验证。

**Comment by innerlee** [comment]
Thanks for the explanation. 

Considering that this change is fairly local (only affects ava dataset), and provides some convinent tool for real use cases, plus @kennymckormick , who will maintain this piece of code, has approved, I will merge it.

There are some alternative ways, and may leave to future,
* Write easy-to-use script tool to transform (taking subsets, remapping labels, etc.) dataset annotations.
* Implement a general purpose class-based dataset weighted-sampling functionality.
